csrc/host_ir/lower.cpp

```diff
@@ -684,6 +687,32 @@ std::unique_ptr<hir::HostIrContainer> HostIrLower::lower(
     hic->addOutput(ir_cloner.clone(output));
   }
 
+  std::vector<Expr*> new_top_level_exprs;
+  for (auto top_level_expr : hic->topLevelExprs()) {
+    if (!isResharding(top_level_expr)) {
+      new_top_level_exprs.push_back(top_level_expr);
+      continue;
+    }
+    for (auto* expr : HostIrLower::lower(top_level_expr, my_device_index)) {
+      // Allocate the recv buffers of communications
+      if (expr->isA<Communication>()) {
+        auto* communication = expr->as<Communication>();
+        TensorView* tv = communication->out();
+        if (tv->getDeviceMesh().has(my_device_index)) {
+          auto* allocate =
+              IrBuilder::create<kir::Allocate>(tv, MemoryType::Global);
+          new_top_level_exprs.push_back(allocate);
+        }
+      }
+      new_top_level_exprs.push_back(expr);
+      if (expr->isA<Communication>()) {
+        auto wait = IrBuilder::create<hir::Wait>(expr->as<Communication>());
+        new_top_level_exprs.push_back(wait);
+      }
+    }
+  }
+  hic->resetTopLevelExprs(new_top_level_exprs);
```

Comment:
Can you clarify why we need reseting? In general, I tend to avoid APIs that reset. They make it hard to reason about life cycles of things. I might even create a new HostIrContainer in order to not having to reset.

Reply:
Instead of reseting, can we add the correct top-level expressions in one go? The high-level logic seems to be: 

```
for each group in topo order: 
  if group leads to a kernel:
    create a PostOnStream and put it in hir's top-level
  else:
    for each expression in topo order:
      if the expression needs lowering:
        lower it to host IR and append the lowered into hir
      else:
        append the expression directly to hir
```

Reply:
In this PR we are separating out the segmentation per-say and lowering a resharding into comms. So there is a first pass in the topological order where we create the top_level_expression, mapping exactly the segments of the SegmentedFusion ; and a second lowering pass that goes through the expression and maybe lower them to (allocation + communication + wait). In a coming PR this second pass will be moved to a different file and become a proper preseg_pass.

Besides being IMO cleaner and more readable this way, it also allows us to control the preseg_pass and their order of execution. This is needed for stream lowering.

For this reason, in this PR, I just separated in two loop what used to be done in "one go" as you are suggesting here. Therefore, I cannot do as you suggest

Reply:
> In a coming PR this second pass will be moved to a different file and become a proper preseg_pass.

Don't bother doing this. Host IR lowering is after segmentation, so I'd rather it being kept in csrc/host_ir not csrc/preseg_passes.

Reply:
I understood the need to change host IR in a container. Many (if not all) host IR optimizations will be implemented as "passes". 

I don't think "reset" is the right API. Most optimizations will be "surgical", e.g., fusing two adjacent for loops and adding stream assignments for overlapping. Such optimizations would "reset" the container with a list of mostly identical host IR besides those being changed. This makes it hard to figure out what's actually changed in the code. I think we'll end up building something like kir::IrVisitor in the future, which is of course not required in this PR.

Reply:
> which segmentation

I understood MultiDeviceExecutor runs two segmentations. FusionExecutorCache is different. It runs only one and host IR lowering is after segmentation (cf. https://docs.google.com/document/d/1QrRmN27XsVjZu7QrZWJJyRENO50878LC3MvlQY1cRYA/edit?tab=t.0)

> should I create another class of Optimization pass for HIR passes

Yes and in a different folder, e.g., csrc/host_ir. This is similar to device_lower/passes. Optimization passes can be as simple as a function, and I'd probably start with just that. When you need to add more features like guards for these passes, you can make them classes.